### ISSUE TITLE
Question about alphago zero evaluator

### ISSUE THREAD

**Issue by Narsil** [gcp/leela-zero#20]
Here is a extract from the original paper.

> Evaluator. To ensure we always generate the best quality data, we evaluate each new neural network checkpoint against the current best network θ∗f before using it for data generation. The neural network θfi is evaluated by the performance of an MCTS search αθi that uses θfi to evaluate leaf positions and prior probabilities (see Search algorithm). Each evaluation consists of 400 games, using an MCTS with 1,600 simulations to select each move, using an infinitesimal temperature τ→ 0 (that is, we deterministically select the move with maximum visit count, to give the strongest possible play). If the new player wins by a margin of > 55% (to avoid selecting on noise alone) then it becomes the best player αθ∗, and is subse-quently used for self-play generation, and also becomes the baseline for subsequent comparisons.

If we use temperature = 0, the selection is deterministic. As I don't believe their is another source of randomness in the play algorithm, all 400 games should always be the same. Am I missing something and only the tested player plays deterministically while the Best model still plays with temperature=1 for the first 30 moves ?
I am trying to write an [implementation](https://github.com/Narsil/alphagozero) to better understand how alphagozero works, and I don't understand how the players will play different games. Do you have any ideas ?

**Comment by gcp** [comment]
There are several sources of randomness, they are all mentioned in the paper. 

They add noise to the policy network priors:
https://github.com/gcp/leela-zero/blob/f321b568ec23d10d024661c6227ddc393a5d8ac8/src/UCTNode.cpp#L181

All network evaluations are done with a random symmetry. This is RANDOM_ROT in Leela's source. 

In practice one expects some randomness from multiprocessing with virtual loss as well.

**Comment by Narsil** [comment]
Isn't the noise only used in self play and not in evaluation ?

I will add the symmetries and rotations, I forgot about those.

**Comment by gcp** [comment]
They only talk setting the softmax temperature (really: disabling the move selection randomization), not about disabling the Dirichlet noise. I suspect you can get away with that anyway, the symmetries add a lot of variation to what it will play.
